### ISSUE TITLE
running timer

### ISSUE THREAD

**Issue by nmrcardoso** [lattice/quda#344]
Running the gauge_alg_test in develop branch i'm getting the error:
ERROR: Cannot start an already running timer (rank 0, host hybrid, ../include/quda_internal.h:127 in Start())
       last kernel called was (name=N4quda9GaugePlaqIfNS_11FloatNOrderIfLi18ELi2ELi18EEEEE,volume=28x28x28x28,aux=threads=614656,prec=4)
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0

**Comment by mikeaclark** [comment]
I'll take a look in a couple of hours.

**Comment by nmrcardoso** [comment]
The problem is that I use TimeProfile to measure time in
pgauge_det_trace.cu
pgauge_heatbath.cu
gauge_fix_ovr.cu
gauge_fix_fft.cu
and it seems that calling one TimeProfile after another one had started is not allowed.
Probably the best way is if I replace all these in the files above by Timer defined in quda_internal?
Please, let me know if this change is ok, replace the Timer, and I'll create a hotfix branch for this.

**Comment by mikeaclark** [comment]
I'm not sure I follow.  Calling one TimeProfile's start while a *different* is already running should be fine.  This error is saying that you're trying to call start on a timer instance that is already running.  

A couple of other observations:
* I see you're using a single TimeProfile called `profileGaugeFix` in `computeGaugeFixingOVRQuda` and `computeGaugeFixingFFTQuda`.  I'd suggest using two different TimeProfile instances for these so they are printed separately.
* Also, the name should end in Quda, not Cuda, so use `GaugeFixFFTQuda` and `GaugeFixOVRQuda`
* The gauge field ordering is hard-coded to MILC_GAUGE_ORDER, this should be changed to use the order set in the `QudaGaugeParam` to allow interfacing with other applications (lines 5307 and 5432 of interface_quda.cpp).

**Comment by nmrcardoso** [comment]
Thanks @mikeaclark .
I think I found the problem in quda_internal.h:
There are 2 repeated lines, see lines 263-266.
    void Start(QudaProfileType idx) { 
      // if total timer isn't running, then start it running
      if (!profile[QUDA_PROFILE_TOTAL].running && idx != QUDA_PROFILE_TOTAL) {
	profile[QUDA_PROFILE_TOTAL].Start(); 
	switchOff = true;
        profile[QUDA_PROFILE_TOTAL].Start();
        switchOff = true;
      }
(...)
the second call triggers the error in:
    void Timer::Start() {
      if (running) errorQuda("Cannot start an already running timer");
      gettimeofday(&start, NULL);
      running = true;
    }

**Comment by mikeaclark** [comment]
Yes that does look like a bug.  I guess that creeped in there with #342.  Can you make a hotfix branch to fix this, and the issues I pointed out above?

**Comment by nmrcardoso** [comment]
I have done the fix on TimeProfile::Start(), and the observations made by @mikeaclark.
Also, I commented the line:
bool tuneSharedBytes() const { return false; } // Don't tune shared memory
in several files.

Still there is a bug in the TimeProfile that I yet not found the origin, since this error is
only triggered in MILC+QUDA.
@mathiaswagner can you look to the code in TimeProfile?

The fixes that I made are on the hotfix/dev_timerprofile branc.

**Comment by mikeaclark** [comment]
@nmrcardoso Can you describe this problem?  What is happening?

Any performance improvement from enabling the shared-memory tuning?

**Comment by nmrcardoso** [comment]
This last problem is related with the problem that I mentioned before, more than one instance is not allowed as in the following example:
  TimeProfile abc( "abcd");
  abc.Start(QUDA_PROFILE_COMPUTE);
(...)
  TimeProfile abcd( "abcd");
  abcd.Start(QUDA_PROFILE_COMPUTE);
(...)
    abc.Stop(QUDA_PROFILE_COMPUTE);
(...)
    abcd.Stop(QUDA_PROFILE_COMPUTE);
(...)
this is related with the global time measurements, for example for measure the total 
QUDA_PROFILE_COMPUTE, in the above example it triggers a double start and double stop call of the global measurement, which returns a Quda error right the second start call of the above example, "Cannot start an already running timer".
Probably the best way is to replace all TimeProfile in the codes except in the interface_quda.cpp?

**Comment by mikeaclark** [comment]
Let's wait for @mathiaswagner to comment on this before we make any big changes.

What you list above should work I think, at least it is supposed to.  The total timer is local to each instance, though all instances share the global timer, but it should be ok to have a double start on these, e.g, 
```c++
    static void StartGlobal(QudaProfileType idx) {
      // if total timer isn't running, then start it running
      if (!global_profile[QUDA_PROFILE_TOTAL].running && idx != QUDA_PROFILE_TOTAL) {
        global_profile[QUDA_PROFILE_TOTAL].Start();
        global_total_level++;
        global_switchOff = true;
      }

      if (idx==QUDA_PROFILE_TOTAL){
        if (global_total_level ==0) global_profile[idx].Start();
        global_total_level++;
      }
      else
        global_profile[idx].Start();
    }
```

**Comment by nmrcardoso** [comment]
That shows only that several calls for QUDA_PROFILE_TOTAL is allowed, but not for 
QUDA_PROFILE_COMPUTE, which is the case here.
And probably the best way is not to use.

Ok, it's better to @mathiaswagner comment on this.

Meanwhile I changed in the interface_quda.cpp to bypass this error by setting the use_global to false in TimeProfile. Since the gauge_fix_ovr.cu and gauge_fix_fft.cu have a TimeProfile for QUDA_PROFILE_COMPUTE, this will not mess with the global timers,
static TimeProfile GaugeFixFFTQuda("GaugeFixFFTQuda",false);
static TimeProfile GaugeFixOVRQuda("GaugeFixOVRQuda",false);

All the changes are in the hotfix/dev_timerprofile branch.
You can merge this into develop if you don't want to have some code broken in develop.




ok, that's better.
Just a new
For now, there is a simple workaround for now, is set false in one of them, to not take into account in the global timers.

**Comment by mikeaclark** [comment]
@nmrcardoso take a look at my branch hotfix/dev_timerprofile2.  I've added tracking information so that when a Timer fails because it is already running or fails because it hasn't been started it should print exactly where this call was made.  Similar to what we do with device_malloc, etc.  

Can you take a look and try to run it with `use_global = true`, to see if it accurately pinpoints where the bad invocation is?  I thought this would help with debugging the timers going forwards.

**Comment by nmrcardoso** [comment]
This is the error using use_global = true in interface_quda.cpp:
ERROR: Cannot start an already running timer (gauge_fix_ovr.cu:1474 in gaugefixingOVR())
ERROR: Aborting (rank 0, host hybrid, ../include/quda_internal.h:129 in Start())
       last kernel called was (name=,volume=,aux=)
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0

code in that line: profileInternalGaugeFixOVR.Start(QUDA_PROFILE_COMPUTE);
this this the second Timer start() with QUDA_PROFILE_COMPUTE type.

There was a compiling problem with this branch with 
mpicxx      -Wall -O3 -D__COMPUTE_CAPABILITY__=350 -DMULTI_GPU -DGPU_GAUGE_ALG -DGPU_GAUGE_TOOLS -DGPU_UNITARIZE -DDIRECT_ACCESS_BLAS -DFERMI_NO_DBLE_TEX -DBUILD_MILC_INTERFACE -DNUMA_AFFINITY -I/usr/local/cuda-7.0/include -DMPI_COMMS  -I/include -I../include -I.  invert_test.cpp -c -o invert_test.o
In file included from /opt/intel//impi/5.0.3.048/intel64/include/mpi.h:2275:0,
                 from invert_test.cpp:20:
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1846:39: error: macro "Start" passed 4 arguments, but takes just 1
                     void *extra_state );
                                       ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1888:49: error: macro "Start" passed 2 arguments, but takes just 1
     virtual void Start( const Group &v1, int v2 ) const
                                                 ^
In file included from ../include/face_quda.h:5:0,
                 from invert_test.cpp:15:
../include/quda_internal.h:301:27: error: expected identifier before ‘__func__’
 #define Start(idx) Start_(__func__, __FILE__, __LINE__, idx)
                           ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1195:18: note: in expansion of macro ‘Start’
     virtual void Start( void ) 
                  ^
../include/quda_internal.h:301:27: error: expected ‘,’ or ‘...’ before ‘__func__’
 #define Start(idx) Start_(__func__, __FILE__, __LINE__, idx)
                           ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1195:18: note: in expansion of macro ‘Start’
     virtual void Start( void ) 
                  ^
In file included from /opt/intel//impi/5.0.3.048/intel64/include/mpi.h:2275:0,
                 from invert_test.cpp:20:
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1843:14: error: field ‘Start’ has incomplete type
     Grequest Start( Query_function  *query_fn,
              ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1888:18: error: variable or field ‘Start’ declared void
     virtual void Start( const Group &v1, int v2 ) const
                  ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1888:18: error: expected ‘;’ at end of member declaration
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1889:5: error: expected unqualified-id before ‘{’ token
     {
     ^
invert_test.cpp:591:1: error: expected ‘}’ at end of input
 }
 ^
In file included from /opt/intel//impi/5.0.3.048/intel64/include/mpi.h:2275:0,
                 from invert_test.cpp:20:
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h: In member function ‘virtual MPI::Group MPI::Win::Get_group() const’:
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:154:44: error: ‘const class MPI::Win’ has no member named ‘Call_errhandler’
     { int err = fnc; if (err) { (_objptr)->Call_errhandler( err ); }}
                                            ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1881:9: note: in expansion of macro ‘MPIX_CALLREF’
         MPIX_CALLREF( this, MPI_Win_get_group( (MPI_Win) the_real_win, &(v2.the_real_group) ));
         ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h: In member function ‘virtual void MPI::Win::Fence(int) const’:
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:154:44: error: ‘const class MPI::Win’ has no member named ‘Call_errhandler’
     { int err = fnc; if (err) { (_objptr)->Call_errhandler( err ); }}
                                            ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1886:9: note: in expansion of macro ‘MPIX_CALLREF’
         MPIX_CALLREF( this, MPI_Win_fence( v1, (MPI_Win) the_real_win ));
         ^
In file included from /opt/intel//impi/5.0.3.048/intel64/include/mpi.h:2275:0,
                 from invert_test.cpp:20:
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h: At global scope:
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1887:5: error: expected unqualified-id at end of input
     }
     ^
/opt/intel//impi/5.0.3.048/intel64/include/mpicxx.h:1887:5: error: expected ‘}’ at end of input
make[1]: *** [invert_test.o] Error 1

**Comment by mikeaclark** [comment]
Ok, thanks for the feedback.  I'll make it more robust (problem is I've made `Start` / `Stop` a macro definition which makes it very flaky and will interfere with other functions with these names.  Once I've cleaned this up I'll merge it into your branch.

**Comment by nmrcardoso** [comment]
Thanks a lot @mikeaclark and sorry for all this work. I was silly doing this kind of stuff, double calls... gah

**Comment by mikeaclark** [comment]
@nmrcardoso, I think it would be better if you switch off the global timer in the algorithms themselves, e.g., gauge_fix_ovr.cu.  This keeps it consistent with the rest of the interface then, so that the global timer represents time spent in the interface and not functions called from there.  Make sense?

**Comment by nmrcardoso** [comment]
yes, I'm doing the changes in the hotfix/dev_timerprofile

**Comment by mikeaclark** [comment]
Ok, I've merged in my branch into yours.  There are now macros TPSTART and TPSTOP that should be called to start/stop a TimeProfile instance.  E.g.
```c++
TimeProfile abc( "abcd");
abc.TPSTART(QUDA_PROFILE_COMPUTE);
abc.TPSTOP(QUDA_PROFILE_COMPUTE);
```

**Comment by mikeaclark** [comment]
Great, let me know when you've finished this, and I will review it all and merge into devel.  Perhaps you can test me pull #345 with your gauge-fixing routines and do this merge? (@mathiaswagner  is at the beach this week....)

**Comment by nmrcardoso** [comment]
done.
Ok, I will test the #345. Just to avoid any bad thing since my first time to do a merge, It's only a click on "merge pull request"?

**Comment by mikeaclark** [comment]
Yes, it's just a case of clicking on merge pull request, and then delete the old branch (also only a click away).  But I would first ask that you checkout this branch and run your gauge fixing tests in it to check all is ok.  It would be nice to know also if you see any performance improvements :)

**Comment by nmrcardoso** [comment]
I finish the checking and left the performance numbers on #345

**Comment by nmrcardoso** [comment]
I've done the merge and deleted the branch #345.
Tomorrow is my last day in work, probably I won't be able to do much things.
However, If I have time I would like to implement the 2D cub reduction, after pulling this hotfix, on the 
pgauge_det_trace.cu
gauge_fix_ovr.cu (gauge fix quality measurement kernel)
gauge_fix_fft.cu (gauge fix quality measurement kernel)

**Comment by mikeaclark** [comment]
Ok, I hope your last day goes ok.  Good luck with the move.

On the writing of the full field necessary for the FFT gauge fixing.  I realised while out on my run that if you bring the parity into the same block, like I did with the plaquette code, then all data necessary to do full coalesced writes for the delta array will be in the same thread block.  So while it would be a pain to get working, i.e., using shared memory exchange to get consecutive threads writing to consecutive memory locations, it is possible.
